File: src/core/identity.ts
// Canonical Core — identity.
//
// Ed25519 over native Web Crypto. No dependencies, runs in Node / Deno / Workers
// / the browser. An identity IS an origin: a namespace and a public key. It
// confers NO trust by itself — trust is not based on origin (see ./cors.ts).

import { b64url, fromB64url, canonicalBytes, type Bytes } from "./codec.ts";

const ALG = { name: "Ed25519" } as const;

export interface Identity {
  /** base64url(raw 32-byte Ed25519 public key) — the origin. */
  readonly id: string;
}

export interface Signer extends Identity {
  sign(data: Bytes): Promise<string>;
}

/** Mint a fresh identity. The private key never leaves the returned closure. */
export async function generateSigner(): Promise<Signer> {
  const kp = (await crypto.subtle.generateKey(ALG, true, ["sign", "verify"])) as CryptoKeyPair;
  const raw = new Uint8Array(await crypto.subtle.exportKey("raw", kp.publicKey));
  const id = b64url(raw);
  return {
    id,
    async sign(data: Bytes): Promise<string> {
      const sig = await crypto.subtle.sign(ALG, kp.privateKey, data);
      return b64url(new Uint8Array(sig));
    },
  };
}

const pubCache = new Map<string, CryptoKey>();

async function importPublic(id: string): Promise<CryptoKey> {
  const cached = pubCache.get(id);
  if (cached) return cached;
  const key = await crypto.subtle.importKey("raw", fromB64url(id), ALG, true, ["verify"]);
  pubCache.set(id, key);
  return key;
}

export async function verifySignature(signerId: string, data: Bytes, sig: string): Promise<boolean> {
  try {
    const key = await importPublic(signerId);
    return await crypto.subtle.verify(ALG, key, fromB64url(sig), data);
  } catch {
    return false;
  }
}

// A signed record carries its signature in `sig`, over the canonical bytes of
// the rest of the object. This is how every primitive in the core is signed.
export async function signRecord<T extends object>(signer: Signer, body: T): Promise<T & { sig: string }> {
  const sig = await signer.sign(canonicalBytes(body));
  return { ...body, sig };
}

export async function verifyRecord<T extends { sig: string }>(signerId: string, record: T): Promise<boolean> {
  const { sig, ...body } = record as { sig: string } & Record<string, unknown>;
  return verifySignature(signerId, canonicalBytes(body), sig);
}

File: crates/canonical-core/src/identity.rs
//! Identity: an Ed25519 keypair. An identity IS an origin — a namespace and a
//! public key. It confers no trust by itself (see cors.rs).

use crate::codec::{b64url, b64url_decode, canonical_string, Json};
use crate::ed25519;
use crate::sys;

/// A public identity — its id is base64url(raw 32-byte Ed25519 public key).
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Identity {
    pub id: String,
}

/// A signing identity. The 32-byte seed never leaves this struct's owner.
#[derive(Clone)]
pub struct Signer {
    seed: [u8; 32],
    pub id: String,
}

impl Signer {
    pub fn generate() -> std::io::Result<Signer> {
        let mut seed = [0u8; 32];
        sys::fill_random(&mut seed)?;
        Ok(Signer::from_seed(seed))
    }

    pub fn from_seed(seed: [u8; 32]) -> Signer {
        let pk = ed25519::public_key(&seed);
        Signer {
            seed,
            id: b64url(&pk),
        }
    }

    pub fn identity(&self) -> Identity {
        Identity {
            id: self.id.clone(),
        }
    }

    pub fn sign(&self, data: &[u8]) -> String {
        b64url(&ed25519::sign(&self.seed, data))
    }

    /// Sign the canonical bytes of a record body (the body excludes `sig`).
    pub fn sign_record(&self, body: &Json) -> String {
        self.sign(canonical_string(body).as_bytes())
    }
}

pub fn verify_signature(id: &str, data: &[u8], sig: &str) -> bool {
    let pk: [u8; 32] = match b64url_decode(id) {
        Some(v) if v.len() == 32 => v.try_into().unwrap(),
        _ => return false,
    };
    let s: [u8; 64] = match b64url_decode(sig) {
        Some(v) if v.len() == 64 => v.try_into().unwrap(),
        _ => return false,
    };
    ed25519::verify(&pk, data, &s)
}

pub fn verify_record(id: &str, body: &Json, sig: &str) -> bool {
    verify_signature(id, canonical_string(body).as_bytes(), sig)
}
